src/librustc/middle/ty/relate.rs

```diff
@@ -13,15 +13,105 @@
 //! can be other things. Examples of type relations are subtyping,
 //! type equality, etc.
 
+use std::cell::Cell;
+use std::iter::FromIterator;
+
 use middle::def_id::DefId;
 use middle::subst::{ErasedRegions, NonerasedRegions, ParamSpace, Substs};
+use middle::traits;
 use middle::ty::{self, Ty, TypeFoldable};
 use middle::ty::error::{ExpectedFound, TypeError};
 use std::rc::Rc;
 use syntax::abi;
 use rustc_front::hir as ast;
 
-pub type RelateResult<'tcx, T> = Result<T, TypeError<'tcx>>;
+
+#[derive(Clone, Debug)]
+pub struct RelateOk<'tcx, T> {
+    pub value: T,
+    pub obligations: Vec<traits::PredicateObligation<'tcx>>,
+}
+impl<'tcx, T> From<T> for RelateOk<'tcx, T> {
+    fn from(value: T) -> Self {
+        RelateOk { value: value, obligations: Vec::new() }
+    }
+}
+impl<'tcx, T> RelateOk<'tcx, T> {
+    pub fn chain<U>(self, other: RelateOk<'tcx, U>) -> RelateOk<'tcx, (T, U)> {
+        let RelateOk { value, mut obligations } = self;
+        let RelateOk { value: other_value, obligations: other_obligations } = other;
+        obligations.extend(other_obligations);
+        RelateOk { value: (value, other_value), obligations: obligations }
+    }
+}
+
+pub type RelateResult<'tcx, T> = Result<RelateOk<'tcx, T>, TypeError<'tcx>>;
+
+pub trait RelateResultTrait<'tcx, T> {
+    fn map_value<U, F: FnOnce(T) -> U>(self, mapper: F) -> RelateResult<'tcx, U>;
+    fn and_then_with<U, F: FnOnce(T) -> RelateResult<'tcx, U>>(self, mapper: F)
+        -> RelateResult<'tcx, U>;
+}
+impl<'tcx, T> RelateResultTrait<'tcx, T> for RelateResult<'tcx, T> {
+    fn map_value<U, F: FnOnce(T) -> U>(self, mapper: F) -> RelateResult<'tcx, U> {
+        match self {
+            Ok(RelateOk { value, obligations }) =>
+                Ok(RelateOk{ value: mapper(value), obligations: obligations }),
+            Err(x) => Err(x),
+        }
+    }
+    fn and_then_with<U, F: FnOnce(T) -> RelateResult<'tcx, U>>(self, mapper: F)
+        -> RelateResult<'tcx, U>
+    {
+        match self {
+            Ok(RelateOk { value, mut obligations }) => match mapper(value) {
+                Ok(RelateOk { value: next_value, obligations: new_obligations }) => {
+                    obligations.extend(new_obligations);
+                    Ok(RelateOk { value: next_value, obligations: obligations })
+                },
+                Err(x) => Err(x),
+            },
+            Err(x) => Err(x),
+        }
+    }
+}
+fn relate_result_from_iter<'tcx, T, U, I: IntoIterator<Item=RelateResult<'tcx, U>>>(iter: I)
+    -> RelateResult<'tcx, T> where T: FromIterator<U>
+{
+    let mut all_obligations = Vec::new();
+    let mut err = None;
+    let end_on_next_elem = Cell::new(false);
+    let value = iter.into_iter()
+        .take_while(|x| {
+            if end_on_next_elem.get() {
```

Comment:
Also, can't we make this a `FromIterator` impl, so that callers only have to write `collect`?

Example: https://github.com/rust-lang/rust/blob/master/src/libcore/result.rs#L889